File: tcpduplex/conn.go
package tcpduplex

import (
	"context"
	"io"
	"net"
	"sync"
	"sync/atomic"

	"github.com/hdmain/tcpduplex/crypto"
	"github.com/hdmain/tcpduplex/protocol"
)

type outbound struct {
	typ uint8
	pt  []byte
}

// Conn is an encrypted full-duplex tcpduplex session after handshake.
// Writes are serialized through a dedicated writer goroutine; reads dispatch either to Receive
// or—when configured—to OnMessage without blocking the record reader beyond bounded channel ops.
type Conn struct {
	nc   net.Conn
	sess *crypto.Session
	cfg  frozenConfig

	sendCh chan outbound

	recvMu     sync.Mutex
	recvErr    error
	recvClosed bool
	recvCh     chan []byte

	msgJobs       chan []byte
	deliverCancel context.CancelFunc
	deliverWG     sync.WaitGroup

	callbackDropped atomic.Uint64

	stopSend chan struct{}

	shutdownStarted atomic.Bool

	writeDone chan struct{}
	readDone  chan struct{}

	closed atomic.Bool
}

func newConn(nc net.Conn, sess *crypto.Session, cfg frozenConfig) *Conn {
	c := &Conn{
		nc:        nc,
		sess:      sess,
		cfg:       cfg,
		sendCh:    make(chan outbound, cfg.SendQueueDepth),
		recvCh:    make(chan []byte, cfg.ReceiveQueueDepth),
		stopSend:  make(chan struct{}),
		writeDone: make(chan struct{}),
		readDone:  make(chan struct{}),
	}
	if cfg.OnMessage != nil {
		c.msgJobs = make(chan []byte, cfg.OnMessageBufferDepth)
		dctx, cancel := context.WithCancel(context.Background())
		c.deliverCancel = cancel
		c.deliverWG.Add(1)
		go c.deliverLoop(dctx)
	}
	go c.readLoop()
	go c.writeLoop()
	return c
}

func (c *Conn) deliverLoop(ctx context.Context) {
	defer c.deliverWG.Done()
	fn := c.cfg.OnMessage
	for {
		select {
		case payload := <-c.msgJobs:
			if fn != nil {
				fn(payload)
			}
		case <-ctx.Done():
			for {
				select {
				case payload := <-c.msgJobs:
					if fn != nil {
						fn(payload)
					}
				default:
					return
				}
			}
		}
	}
}

// Send queues an application message (encrypted as MsgText).
func (c *Conn) Send(payload []byte) error {
	return c.SendContext(context.Background(), payload)
}

// SendContext waits until payload is accepted into the outbound queue or ctx is done.
func (c *Conn) SendContext(ctx context.Context, payload []byte) error {
	if c.closed.Load() {
		return ErrClosed
	}
	if len(payload) > c.cfg.MaxMessageBytes {
		return ErrMessageTooLarge
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	body := append([]byte(nil), payload...)
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-c.stopSend:
		return ErrClosed
	case c.sendCh <- outbound{typ: protocol.MsgText, pt: body}:
		return nil
	}
}

// Receive blocks until the next application message or an error (requires OnMessage == nil).
func (c *Conn) Receive() ([]byte, error) {
	return c.ReceiveContext(context.Background())
}

// ReceiveContext waits for the next MsgText or termination subject to ctx.
func (c *Conn) ReceiveContext(ctx context.Context) ([]byte, error) {
	if c.cfg.OnMessage != nil {
		return nil, ErrReceiveDisabled
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case msg, ok := <-c.recvCh:
		if !ok {
			c.recvMu.Lock()
			err := c.recvErr
			c.recvMu.Unlock()
			if err == nil {
				err = ErrClosed
			}
			return nil, err
		}
		return msg, nil
	}
}

// CallbackDropped counts inbound messages dropped because OnMessageBufferDepth was exhausted.
func (c *Conn) CallbackDropped() uint64 {
	return c.callbackDropped.Load()
}

// Close performs graceful shutdown with unlimited waits (writes flush, close record, reads drain).
func (c *Conn) Close() error {
	return c.Shutdown(context.Background())
}

// Shutdown tears down the session honoring ctx for waits on writer/delivery completion.
func (c *Conn) Shutdown(ctx context.Context) error {
	if ctx == nil {
		ctx = context.Background()
	}
	if !c.shutdownStarted.CompareAndSwap(false, true) {
		return ErrClosed
	}

	c.closed.Store(true)
	close(c.stopSend)

	err := waitChan(ctx, c.writeDone)
	if c.deliverCancel != nil {
		c.deliverCancel()
		err = firstNonNil(err, waitWaitGroup(ctx, &c.deliverWG))
	}

	closeErr := c.nc.Close()
	err = firstNonNil(err, waitChan(ctx, c.readDone))
	if err != nil {
		return err
	}
	return closeErr
}

func waitChan(ctx context.Context, ch <-chan struct{}) error {
	select {
	case <-ch:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func waitWaitGroup(ctx context.Context, wg *sync.WaitGroup) error {
	done := make(chan struct{})
	go func() {
		wg.Wait()
		close(done)
	}()
	select {
	case <-done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func firstNonNil(a, b error) error {
	if a != nil {
		return a
	}
	return b
}

func (c *Conn) abortRecv(err error) {
	c.recvMu.Lock()
	defer c.recvMu.Unlock()
	if c.recvClosed {
		return
	}
	c.recvClosed = true
	if err == nil {
		err = io.EOF
	}
	c.recvErr = err
	close(c.recvCh)
}

func (c *Conn) readLoop() {
	defer close(c.readDone)

	for {
		msgType, sealed, err := protocol.ReadRecord(c.nc)
		if err != nil {
			c.abortRecv(err)
			return
		}
		pt, err := c.sess.Open(sealed)
		if err != nil {
			c.abortRecv(err)
			return
		}
		switch msgType {
		case protocol.MsgText:
			if len(pt) > c.cfg.MaxMessageBytes {
				c.abortRecv(ErrMessageTooLarge)
				return
			}
			if c.cfg.OnMessage != nil {
				payload := append([]byte(nil), pt...)
				select {
				case c.msgJobs <- payload:
				default:
					if c.cfg.DisconnectOnSlowCallbackConsumer {
						c.abortRecv(ErrSlowConsumer)
						return
					}
					c.callbackDropped.Add(1)
				}
			} else {
				select {
				case c.recvCh <- append([]byte(nil), pt...):
				case <-c.stopSend:
					c.abortRecv(ErrClosed)
					return
				}
			}
		case protocol.MsgPing:
			reply := append([]byte(nil), pt...)
			select {
			case <-c.stopSend:
				c.abortRecv(ErrClosed)
				return
			case c.sendCh <- outbound{typ: protocol.MsgPong, pt: reply}:
			}
		case protocol.MsgPong:
		case protocol.MsgClose:
			c.abortRecv(io.EOF)
			return
		default:
			c.abortRecv(protocol.ErrBadFrame)
			return
		}
	}
}

func (c *Conn) writeLoop() {
	defer close(c.writeDone)

	writeOut := func(out outbound) bool {
		sealed, err := c.sess.Seal(out.pt)
		if err != nil {
			return false
		}
		if err := protocol.WriteRecord(c.nc, out.typ, sealed); err != nil {
			return false
		}
		return true
	}

	for {
		select {
		case <-c.stopSend:
			goto flush
		case out := <-c.sendCh:
			if !writeOut(out) {
				return
			}
		}
	}
flush:
	for {
		select {
		case out := <-c.sendCh:
			if !writeOut(out) {
				return
			}
		default:
			_ = writeOut(outbound{typ: protocol.MsgClose, pt: nil})
			return
		}
	}
}

// Underlying returns the TCP connection wrapped by this session.
func (c *Conn) Underlying() net.Conn {
	return c.nc
}

File: lib/src/main/kotlin/com/hdmain/tcpduplex/Conn.kt
package com.hdmain.tcpduplex

import com.hdmain.tcpduplex.crypto.Session
import com.hdmain.tcpduplex.protocol.Protocol
import java.io.EOFException
import java.net.Socket
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference

private data class Outbound(val typ: UByte, val pt: ByteArray)

/**
 * Encrypted full-duplex tcpduplex session after handshake.
 * Writes are serialized through a dedicated writer thread; reads dispatch either to [receive]
 * or—when configured—to [Config.onMessage].
 */
class Conn internal constructor(
    private val socket: Socket,
    private val session: Session,
    private val cfg: FrozenConfig,
) : AutoCloseable {
    private val input = socket.getInputStream()
    private val output = socket.getOutputStream()

    private val sendQueue = ArrayBlockingQueue<Outbound>(cfg.sendQueueDepth)
    private val pendingRecv = ArrayDeque<ByteArray>()
    private val recvLock = Object()
    private val msgJobs: ArrayBlockingQueue<ByteArray>? =
        if (cfg.onMessage != null) ArrayBlockingQueue(cfg.onMessageBufferDepth) else null

    private val recvErr = AtomicReference<Throwable?>(null)
    private val recvClosed = AtomicBoolean(false)
    private val callbackDropped = AtomicLong(0)
    private val stopSend = AtomicBoolean(false)
    private val shutdownStarted = AtomicBoolean(false)
    private val closed = AtomicBoolean(false)

    private val writeDone = CountDownLatch(1)
    private val readDone = CountDownLatch(1)
    private val deliverDone = CountDownLatch(if (cfg.onMessage != null) 1 else 0)

    private val deliverStop = AtomicBoolean(false)

    init {
        if (cfg.onMessage != null) {
            Thread({ deliverLoop() }, "tcpduplex-deliver").apply { isDaemon = true }.start()
        }
        Thread({ readLoop() }, "tcpduplex-read").apply { isDaemon = true }.start()
        Thread({ writeLoop() }, "tcpduplex-write").apply { isDaemon = true }.start()
    }

    private fun deliverLoop() {
        try {
            val fn = cfg.onMessage ?: return
            while (!deliverStop.get()) {
                val payload = msgJobs?.poll(100, TimeUnit.MILLISECONDS) ?: continue
                fn(payload)
            }
            while (true) {
                val payload = msgJobs?.poll() ?: break
                fn(payload)
            }
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        } finally {
            deliverDone.countDown()
        }
    }

    /** Queues an application message (encrypted as MsgText). */
    fun send(payload: ByteArray) {
        send(null, payload)
    }

    /** Waits until [payload] is accepted into the outbound queue or the deadline elapses. */
    fun send(deadlineMillis: Long?, payload: ByteArray) {
        if (closed.get()) throw ClosedException()
        if (payload.size > cfg.maxMessageBytes) throw MessageTooLargeException()
        val body = payload.copyOf()
        val outbound = Outbound(Protocol.MSG_TEXT, body)
        if (deadlineMillis == null) {
            while (true) {
                if (stopSend.get()) throw ClosedException()
                if (sendQueue.offer(outbound)) return
                Thread.sleep(1)
            }
        } else {
            val deadline = System.currentTimeMillis() + deadlineMillis
            while (System.currentTimeMillis() < deadline) {
                if (stopSend.get()) throw ClosedException()
                if (sendQueue.offer(outbound)) return
                Thread.sleep(1)
            }
            throw java.util.concurrent.TimeoutException()
        }
    }

    /** Blocks until the next application message or an error (requires onMessage == null). */
    fun receive(): ByteArray = receive(null)

    /** Waits for the next MsgText or termination subject to an optional deadline. */
    fun receive(deadlineMillis: Long?): ByteArray {
        if (cfg.onMessage != null) throw ReceiveDisabledException()
        val deadlineAt = if (deadlineMillis != null) System.currentTimeMillis() + deadlineMillis else null
        synchronized(recvLock) {
            while (true) {
                recvErr.get()?.let { throw it }
                if (pendingRecv.isNotEmpty()) {
                    return pendingRecv.removeFirst()
                }
                if (recvClosed.get()) {
                    throw recvErr.get() ?: ClosedException()
                }
                if (deadlineAt == null) {
                    recvLock.wait()
                } else {
                    val remaining = deadlineAt - System.currentTimeMillis()
                    if (remaining <= 0) throw java.util.concurrent.TimeoutException()
                    recvLock.wait(remaining)
                }
            }
        }
    }

    /** Counts inbound messages dropped because onMessageBufferDepth was exhausted. */
    fun callbackDropped(): Long = callbackDropped.get()

    override fun close() {
        shutdown(null)
    }

    /** Tears down the session honoring an optional wait deadline. */
    fun shutdown(waitMillis: Long?) {
        if (!shutdownStarted.compareAndSet(false, true)) {
            throw ClosedException()
        }
        closed.set(true)
        stopSend.set(true)

        awaitLatch(writeDone, waitMillis)
        deliverStop.set(true)
        if (cfg.onMessage != null) {
            awaitLatch(deliverDone, waitMillis)
        }

        val closeErr = runCatching { socket.close() }.exceptionOrNull()
        awaitLatch(readDone, waitMillis)
        if (closeErr != null) throw closeErr
    }

    private fun awaitLatch(latch: CountDownLatch, waitMillis: Long?) {
        if (waitMillis == null) {
            latch.await()
        } else if (!latch.await(waitMillis, TimeUnit.MILLISECONDS)) {
            throw java.util.concurrent.TimeoutException()
        }
    }

    private fun abortRecv(err: Throwable?) {
        synchronized(recvLock) {
            if (recvClosed.get()) return
            recvClosed.set(true)
            recvErr.set(err ?: EOFException())
            recvLock.notifyAll()
        }
    }

    private fun enqueueRecv(payload: ByteArray) {
        synchronized(recvLock) {
            while (pendingRecv.size >= cfg.receiveQueueDepth) {
                if (stopSend.get()) {
                    abortRecv(ClosedException())
                    return
                }
                recvLock.wait(10)
            }
            pendingRecv.addLast(payload)
            recvLock.notifyAll()
        }
    }

    private fun readLoop() {
        try {
            while (true) {
                val (msgType, sealed) = try {
                    Protocol.readRecord(input)
                } catch (e: Exception) {
                    abortRecv(e)
                    return
                }
                val pt = try {
                    session.open(sealed)
                } catch (e: Exception) {
                    abortRecv(e)
                    return
                }
                when (msgType) {
                    Protocol.MSG_TEXT -> {
                        if (pt.size > cfg.maxMessageBytes) {
                            abortRecv(MessageTooLargeException())
                            return
                        }
                        if (cfg.onMessage != null) {
                            val payload = pt.copyOf()
                            if (!msgJobs!!.offer(payload)) {
                                if (cfg.disconnectOnSlowCallbackConsumer) {
                                    abortRecv(SlowConsumerException())
                                    return
                                }
                                callbackDropped.incrementAndGet()
                            }
                        } else {
                            if (stopSend.get()) {
                                abortRecv(ClosedException())
                                return
                            }
                            enqueueRecv(pt.copyOf())
                        }
                    }
                    Protocol.MSG_PING -> {
                        val reply = pt.copyOf()
                        while (true) {
                            if (stopSend.get()) {
                                abortRecv(ClosedException())
                                return
                            }
                            if (sendQueue.offer(Outbound(Protocol.MSG_PONG, reply))) break
                            Thread.sleep(1)
                        }
                    }
                    Protocol.MSG_PONG -> Unit
                    Protocol.MSG_CLOSE -> {
                        abortRecv(EOFException())
                        return
                    }
                    else -> {
                        abortRecv(Protocol.BadFrameException())
                        return
                    }
                }
            }
        } finally {
            readDone.countDown()
        }
    }

    private fun writeLoop() {
        try {
            fun writeOut(out: Outbound): Boolean {
                return try {
                    val sealed = session.seal(out.pt)
                    Protocol.writeRecord(output, out.typ, sealed)
                    true
                } catch (_: Exception) {
                    false
                }
            }

            while (!stopSend.get()) {
                val out = sendQueue.poll(100, TimeUnit.MILLISECONDS) ?: continue
                if (!writeOut(out)) return
            }

            // flush remaining outbound messages
            flush@ while (true) {
                val out = sendQueue.poll() ?: break@flush
                if (!writeOut(out)) return
            }
            writeOut(Outbound(Protocol.MSG_CLOSE, ByteArray(0)))
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        } finally {
            writeDone.countDown()
        }
    }

    /** Returns the TCP socket wrapped by this session. */
    fun underlying(): Socket = socket
}
